basic game lobby

Reptudn opened the pull request:
- games can start
- lobby state should be synced between clients
- leaving game works partly (only when clicking on the dedicated game buttons)
- local players dont really do anything yet and are basically useless as of now


when leaving a game not over the buttons you are still in the game cause i havent figured out a good way to handle that yet...

Reptudn commented:
I hope that is is enough for you to enhance the game

FreddyMSchubert commented:
What is the enable powerups functionality for? Does that do something? Isn't the ability to disable or enable every powerup specifically going a bit overboard?

FreddyMSchubert commented:
You can end up on a game screen with updating game data at the bottom but no rendering, why?

FreddyMSchubert commented on app/frontend-scripts/game.ts:
```diff
@@ -4,40 +4,58 @@ import { loadPartialView } from './script.js';
-	playerId
-);
-const wsUrl = `/api/games/join/${gameId}/${playerId}`;
+const wsUrl = `/api/games/connect?gameId=${gameId}`; // TODO: make this url with env var (ps to myself.. this doesnt work in the frontend xd)
```
Can you please go through the TODOs you put in your code and checking whether they're up to date before making the PR

Reptudn replied:
sure master

Reptudn replied:
done

FreddyMSchubert commented on app/frontend-scripts/game_page.ts:
```diff
@@ -0,0 +1,5 @@
+document
```
What does this file do?

Reptudn replied:
back then it was supposed to be shown when on the game page to see all recent games before being able to create a game but thats not there anymore. maybe in future we just add it to the index page of the user when logged in cuz we dont use that page yet

FreddyMSchubert replied:
Please delete it if it doesn't do anything.

FreddyMSchubert commented on app/frontend-scripts/game_setup.ts:
```diff
@@ -1,254 +1,235 @@
 	}
-	playersContainer.removeChild(currentPlayers[currentPlayers.length - 1]);
+
+	onlineFriendsContainer.innerHTML = friends.map((friend) => `
```
Why not do this using ejs?

FreddyMSchubert commented on app/frontend-scripts/game_setup.ts:
```diff
@@ -1,254 +1,235 @@
 
-	console.log('Sending data to server:', data);
-	fetch('/api/games/start', {
+// TODO: doesnt work
```
Is kicking the same as removing? We should have consistent naming here. (And that naming should be removing as they can be readded, right?) For me it doesn't remove a player but the entire game. Starting the game shows the game doesn't exist, doing anything as the admin tells you you're not an admin.

FreddyMSchubert commented on app/frontend-scripts/game_setup.ts:
```diff
@@ -1,254 +1,235 @@
+	showLocalInfo(`${data.message || 'Player kicked successfully!'}`);
+}
+
+export async function leaveGame() {
```
I don't think this is even necessary we can have non-ragequit leaves also handled by the pinging

Reptudn replied:
i think its nice to have for ux.. when we see that the socket was closed without leaving manually we can add the ragequit achievement.

also i think pinging is also not necessary since we should be able to get the error code from our server websocket of the cllient how they disconnected and handle it accordingly

FreddyMSchubert commented on app/frontend-scripts/script.ts:
```diff
@@ -3,13 +3,15 @@ import { showLocalError, showLocalInfo } from './alert.js';
 	interface Window {
 		updateActiveMenu: (selectedPage: string) => void;
-		loadPartialView: (page: string, pushState?: boolean) => Promise<void>;
+		loadPartialView: (page: string, pushState?: boolean, subroute?: string | null, isPartial?: boolean) => Promise<void>;
```
What? It's called loadPartialView 😂. What does this do? Why isn't every page just... partial? Please also add a comment or rename the variable.

Reptudn replied:
done

FreddyMSchubert commented on app/frontend-scripts/script.ts:
```diff
@@ -25,24 +27,68 @@ export function updateActiveMenu(selectedPage: string): void {
 	document.head.title = `Transcendence: ${selectedPage}`;
 }
 
+export async function createGame()
```
Why do it like this? Can't this just be loaded by game setup? For subsequent presses after the js has been loaded maybe we can add a new event listener to the button. The main script is not a good place to put this.

Reptudn replied:
i can put it there sure

Reptudn replied:
no wait i remember why i put it there.. so that when the game fails to create that i dont have to send the user back to the profile page right away cuz that would make the game setup page flash real quick before going back

FreddyMSchubert commented on app/public/css/main.css:
```diff
@@ -13,6 +13,48 @@
 
 @layer utilities {
+
+	.glow-rainbow-border {
```
Does this work yet? How can I try it? I can't see it.

FreddyMSchubert commented on app/public/pages/game_page.ejs:
```diff
@@ -0,0 +1,17 @@
+<div>
```
What does this do and how do I get to it?

`{"statusCode":500,"error":"Internal Server Error","message":"/workspaces/42_transcendence/app/public/pages/game_page.ejs:8\n    6| \t<!-- display recent games here with a list of divs where green bg is won and red bg is lost -->\n    7| \t<div class=\"recent_games\">\n >> 8| \t\t<% if (recent_games.length === 0) { %>\n    9| \t\t\t<p class=\"glow-green\">No recent games found.</p>\n    10| \t\t<%  } else { recent_games.forEach(game => { %>\n    11| \t\t<div class=\"game_card <%= game.winner_id === user_id ? 'bg-green-500' : 'bg-red-500' %> p-4 mb-2 rounded\">\n\nrecent_games is not defined"}`

FreddyMSchubert commented on app/src/routes/api/games/index.ts:
```diff
@@ -1,215 +1,585 @@
+		}
+	);
+
+	// TODO: add invite cancel
```
I don't think that's necessary. Also please remove your todos we shouldn't have them on master

FreddyMSchubert commented on app/src/services/database/games.ts:
```diff
@@ -0,0 +1,186 @@
+	return { id: result.lastID, code: game_code }; // last id is the id of the created game + game_code
+}
+
+export async function setGameStatus(
```
Why is this still how it works? I thought we agreed it was unnecessary to save currently running games in the DB and to only create an entry when it's done?

FreddyMSchubert commented on app/src/services/database/games.ts:
```diff
@@ -0,0 +1,186 @@
+	}
+}
+
+export async function removeUserFromGame(
```
None of this should need to be done

FreddyMSchubert commented on app/src/services/pong/games/gameFormats.ts:
```diff
@@ -0,0 +1,361 @@
+	return defaultBotNames[Math.floor(Math.random() * defaultBotNames.length)];
+}
+
+// TODO: there is one player too much in the game always on the map
```
What?
Furthermore, why does this file have functions at all? Please rename it.

Reptudn replied:
done

FreddyMSchubert commented on app/src/services/pong/games/rawMapHandler.ts:
```diff
@@ -81,3 +83,28 @@ function isMapConditionFulfilled(
 	else return false;
 }
+
+// TODO: doesnt really work yet
```
😂

Reptudn replied:
this is fixed cuz it worked i just didnt remove the todo

FreddyMSchubert commented on app/src/types/Games.d.ts:
```diff
@@ -1,16 +1,27 @@
+interface Powerups{
```
Still confused about the powerups. Can this be edited or made to be editable in some non-tedious manual way?

FreddyMSchubert commented on tsconfig.frontend.json:
```diff
@@ -2,8 +2,8 @@
 		"outDir": "app/public/js",
-		"module": "ESNext",
-		"moduleResolution": "Bundler",
+		"module": "esnext",
```
Hell yeah dude these are the kinda changes I wanna see 🔥

Reptudn replied:
haha

FreddyMSchubert reviewed (commented):
Please:

- [ ] Fix kicking players
- [ ] Handle disconnects gracefully via pinging
- [ ] Remove detailed powerup setting
- [ ] Fix bug where game doesn't always render
- [ ] Only create a db entry for a game after it's done
- [ ] All the stuff I put into the comments

In general I think it would be helpful to clean up the code more and sort it into clear distinct files.
Very cool but still needs improvements to not be annoying on master forever. Very cool thank you 🚀

Reptudn commented:
> What is the enable powerups functionality for? Does that do something? Isn't the ability to disable or enable every powerup specifically going a bit overboard?

I added it because you said you wanted powerups so I just added a toggle already

Reptudn commented:
> Please:
> 
> 
> 
> - [ ] Fix kicking players
> 
> - [ ] Handle disconnects gracefully via pinging
> 
> - [ ] Remove detailed powerup setting
> 
> - [ ] Fix bug where game doesn't always render
> 
> - [ ] Only create a db entry for a game after it's done
> 
> - [ ] All the stuff I put into the comments
> 
> 
> 
> In general I think it would be helpful to clean up the code more and sort it into clear distinct files.
> 
> Very cool but still needs improvements to not be annoying on master forever. Very cool thank you 🚀

I didn't add any end game logic really since i just wanted to create the lobby belle the game and enhance the general game classes but I can create a db entry after the game even though I don't really know what to put there since the game logic is not there that someone can win a game

I don't know if we have to do pinging but I will see since I think websockets throw an error once someone disconnects idk will see

I thought I fixed kicking players.. will take a look into it 👍

FreddyMSchubert commented:
Oh yeah that's alright you don't have to do any DB interactions on this branch at all I think that's all my job.

Reptudn commented:
> Oh yeah that's alright you don't have to do any DB interactions on this branch at all I think that's all my job.

i added some todos where i planned the db stuff to happen

NoelSabia commented:
schmutzcode

Reptudn commented:
> das war nur konstruktiv gemeint!
> 
> 
> 
> > > schmutzcode
> 
> > 
> 
> > Halts Maul
> 
> 
> 
> 

Ok Check
✂️

Reptudn commented:
@FreddyMSchubert i think this should be good enough now to be able to merge it into main that you can work on the game and i can fix other small issues plus creating a tournament system

FreddyMSchubert commented on app/src/services/pong/games/playerClass.ts:
Love it way better

Reptudn replied:
good

FreddyMSchubert reviewed (approved):
Dude the frontend code is a fever dream. I think we gotta go over that in an issue once and clean it up. I am about to cry.
<img width="210" height="84" alt="Screenshot 2025-07-21 at 23 04 00" src="https://github.com/user-attachments/assets/9a2d77ab-9c95-42e7-bf50-dd7e0ebe0b46" />
Why are all the files fully unsorted in a single folder 😭 Many unused functions and random duplicated HTML we have in ejs already. Also just very mid code.
This PR has not made it better but I need it merged so oh well.
